modules/library/render/src/test/java/org/geotools/renderer/lite/StreamingRendererTest.java

```diff
@@ -927,4 +933,136 @@ public void errorOccurred(Exception e) {
+        sr.setMapContent(mc);
+
+        ReferencedEnvelope envelope =
+                new ReferencedEnvelope(new Envelope(-180, 180, -90, 90), WGS84);
+        BufferedImage bi = new BufferedImage(100, 100, BufferedImage.TYPE_3BYTE_BGR);
+        Graphics2D graphics = bi.createGraphics();
+        try {
+            sr.paint(graphics, new Rectangle(5, 5, 7, 7), envelope);
+        } finally {
+            graphics.dispose();
+            mc.dispose();
+        }
+
+        assertEquals(1, properties.size());
+        assertArrayEquals(
+                new String[] {"/complex/props/color", "geom", "complex"}, properties.get(0));
+    }
+
+    /**
+     * A minimalistic MockPropertyAccessor (and its factory), which mocks xml xpath evaluation. It
+     * is registered through the
+     * META-INF/services/org.geotools.filter.expression.PropertyAccessorFactory SPI registration in
+     * the test/resources.
+     */
+    public static class MockPropertyAccessorFactory implements PropertyAccessorFactory {
+        @Override
+        public PropertyAccessor createPropertyAccessor(
+                Class<?> type, String xpath, Class<?> target, Hints hints) {
+            if (!SimpleFeature.class.isAssignableFrom(type)
+                    && !SimpleFeatureType.class.isAssignableFrom(type)) {
+                return null;
+            }
+            return new PropertyAccessor() {
+                @Override
+                public boolean canHandle(Object object, String xpath, Class<?> target) {
+                    String root = getRoot(xpath);
+                    if (root == null) {
+                        return false;
+                    }
+                    AttributeDescriptor descriptor = null;
+                    if (object instanceof SimpleFeature) {
+                        descriptor = ((SimpleFeature) object).getType().getDescriptor(root);
+                    } else if (object instanceof SimpleFeatureType) {
+                        descriptor = ((SimpleFeatureType) object).getDescriptor(root);
+                    }
+                    return descriptor != null
+                            && Boolean.TRUE.equals(descriptor.getUserData().get("xml"));
+                }
+
+                @Override
+                public <T> T get(Object object, String xpath, Class<T> target)
+                        throws IllegalArgumentException {
+                    String root = getRoot(xpath);
+                    if (root == null) {
+                        throw new IllegalArgumentException("Invalid xpath: " + xpath);
+                    }
+                    if (object instanceof SimpleFeature) {
+                        String xml = (String) ((SimpleFeature) object).getAttribute(root);
+                        // we don't care about actually extracting the xpath value in our test
+                        return (T) xml;
+                    } else if (object instanceof SimpleFeatureType) {
```

Comment:
@jodygarnett I'd like to hear some opinion about this bit. Is it correct for a property accessor to return the descriptor of a property that does not match what is actually being accessed, but a xpath parent of it?

Reply:
If I am reading this code correctly `root` is the the initial attribute name in the xpath,
so this is going to return one of the property descriptors from the SimpleFeatureType. Should the the property descriptor matching the value returned by the "instance SimpleFeature" case above...

Where do you see " property that does not match what is actually being accessed, but a xpath parent of it?"

Reply:
The path in the test is "/complex/props/color", the attribute descriptor returned is for the "/complex" property, because the "props/color" part is in the XML contained in the "complex" property, which is a string. So the path is pointing to one attribute, but the descriptor returned is not for it, it's for the "parent's parent" of it.

Reply:
Oh I see, what a pain.

To be honest:
- each property accessor was supposed to access one thing (complex feature, simple feature type, java.util.map, java bean, json object, whatever)
- we had an xpath library (JXPath?) would use property accessors to drill down through this stuff
- Just added the ability to use XPath to drill into FeatureType, SimpleFeatureType, ComplexFeatureType ... and I kind of went with it.

So if I have the design correct, the XPath should drill into the type system should drill down as far as it is able.

In this case it should return a PropertyDescriptor for color binding to Color.class etc...

Reply:
Right... so it would need to have a schema of the XML in question and look it up, or just give up and assume everything is a string (the little XML snipped in the test is "<props><color>blue</color></props>", which does not carry any info).
@elahrvivaz I guess this means going the "function way" is probably better.

Reply:
Thanks, I will close the PR. I think our approach was designed to get around having a well-defined schema, to allow schema-less json blobs, since e.g. there would not be a property descriptor for color in the feature type. seems to have worked surprisingly well until hitting this issue :)